src/oidc_child/oidc_child_json.c

```diff
@@ -0,0 +1,477 @@
+        } else {
+            str = talloc_asprintf_append(str, "%%20%s", json_string_value(s));
+        }
+        if (str == NULL) {
+            DEBUG(SSSDBG_OP_FAILURE, "Failed to copy '%s' string.\n", attr);
+            return NULL;
+        }
+
+    }
+
+    return str;
+}
+
+static errno_t get_endpoints(json_t *inp, struct devicecode_ctx *dc_ctx)
+{
+    int ret;
+
+    dc_ctx->device_authorization_endpoint = get_json_string(dc_ctx, inp,
+                                               "device_authorization_endpoint");
+    if (dc_ctx->device_authorization_endpoint == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Missing device_authorization_endpoint in "
+                                   "openid configuration.\n");
+        ret = EINVAL;
+        goto done;
+    }
+    dc_ctx->token_endpoint = get_json_string(dc_ctx, inp, "token_endpoint");
+    if (dc_ctx->token_endpoint == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Missing token_endpoint in openid "
+                                   "configuration.\n");
+        ret = EINVAL;
+        goto done;
+    }
+    dc_ctx->userinfo_endpoint = get_json_string(dc_ctx, inp,
+                                                "userinfo_endpoint");
+    if (dc_ctx->userinfo_endpoint == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Missing userinfo_endpoint in openid "
+                                   "configuration.\n");
+        ret = EINVAL;
+        goto done;
+    }
+
+    dc_ctx->jwks_uri = get_json_string(dc_ctx, inp, "jwks_uri");
+    if (dc_ctx->jwks_uri == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Missing jwks_uri in openid "
+                                   "configuration.\n");
+    }
+
+    dc_ctx->scope = get_json_scope(dc_ctx, inp, "scopes_supported");
+    if (dc_ctx->scope == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Missing scopes in openid "
+                                   "configuration.\n");
+    }
+
+    ret = EOK;
+done:
+    return ret;
+}
+
+static errno_t str_to_jws(TALLOC_CTX *mem_ctx, const char *inp, json_t **jws)
+{
+    char *pl;
+    char *sig;
+    json_t *o;
+    int ret;
+    char *str;
+
+    str = talloc_strdup(mem_ctx, inp);
+    if (str == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to copy token string.\n");
+        return ENOMEM;
+    }
+
+    pl = strchr(str, '.');
+    if (pl == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "String does not look like serialized JWS, missing first '.'\n");
+        talloc_free(str);
+        return EINVAL;
+    }
+    *pl = '\0';
+    pl++;
+
+    sig = strchr(pl, '.');
+    if (sig == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "String does not look like serialized JWS, missing second '.'\n");
+        talloc_free(str);
+        return EINVAL;
+    }
+    *sig = '\0';
+    sig++;
+
+    o = json_object();
+    if (o == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to create JSON object.\n");
+        return EIO;
```

Comment:
`str` leaks here

Reply:
Probably it would be easier to simply move it to the beginning of the function.

Reply:
Ok, I switched to a single return at the end and a done block.